### PR TITLE
[14.0][ADD]sale_order_general_discount_payment_term: Add sale order general discount on payment term

### PR DESCRIPTION
**Pull request by odooNextev**
This modulo allow to add sale_order general discount based on payment term discount field.
Also hide discount from partner to avoid conflicts

### CODE REVIEW

**Review by francesco-ooops — APPROVED** [review]
Functional ok!

File: sale_order_general_discount_payment_term/models/account_payment_term.py
Comment (HekkiMelody) [inline sale_order_general_discount_payment_term/models/account_payment_term.py]: No point restating obvious defaults IMO. (This is a new field, not an override, correct?)

Instead, I think it'd be good to add the same modifiers that the corresponding field has in `sale_order_general_discount`

```py
    sale_discount = fields.Float(
        digits=dp.get_precision('Discount'),
        string='Discount (%)',
        company_dependent=True,
        track_visibility="onchange",
    )
```

File: sale_order_general_discount_payment_term/views/res_partner.xml
Comment (HekkiMelody) [inline sale_order_general_discount_payment_term/views/res_partner.xml]:
```suggestion
            <field name="sale_discount" position="attributes">
                <attribute name="invisible">1</attribute>
            </field>
```

is this not sufficient, or does it not work for some reason?

File: sale_order_general_discount_payment_term/models/sale_order.py
Comment (HekkiMelody) [inline sale_order_general_discount_payment_term/models/sale_order.py]: `api.depends` and `_compute` functions are generally used if they're computing the value of a field.

I think this should be an `api.onchange`

Cf. the module `sale_order_general_discount`:

```py
    @api.onchange('partner_id')
    def onchange_partner_id(self):
        super().onchange_partner_id()
        self.general_discount = self.partner_id.sale_discount
        return
```

**Review by HekkiMelody — CHANGES_REQUESTED** [review]
You hide the original field, but for existing installations the value of `sale_discount` on `res.partner`s will still exist and it will keep being applied to `sale.order`s, with no way for the user to change it.

There are a few options to deactivate it entirely, One of them could be a `post_init_hook` that just writes `0` on every partner's `sale_discount` field. I don't know if people would find it too aggressive, though.

Another option is to override `onchange_partner_id`, save the existing discount, call `super()` and then restore the discount that existed before the `onchange` was called, so as to nullify the effect of the change in `sale_order_general_discount` (every line's discount would be recalculated twice for nothing, but alas, I've seen worse in Odoo :D ).

Alternatively, since `sale_order_general_discount` is so small and you want to deactivate half of it, you could just replicate its functionality and not depend on it. (and maybe add it to the `excludes` key in the manifest so people don't accidentally install both).

I think option number 2 (overriding `onchange_partner_id` is by far the most preferable, though)

File: sale_order_general_discount_payment_term/views/res_partner.xml
Comment (odooNextev) [reply sale_order_general_discount_payment_term/views/res_partner.xml]: Hi @aleuffre, thanks for your review. Set "invisible" on field sale_discount it doesn't works properly because we need to hide a div with sale_discount and a string inside.

File: sale_order_general_discount_payment_term/models/sale_order.py
Comment (odooNextev) [reply sale_order_general_discount_payment_term/models/sale_order.py]: Hi @aleuffre, I think that is 12.0 version. In v14 this is the code in sale_order_general_discount:

@api.depends("partner_id")
    def _compute_general_discount(self):
        for so in self:
            so.general_discount = so.partner_id.sale_discount

Comment (HekkiMelody) [reply sale_order_general_discount_payment_term/models/sale_order.py]: Ah, my bad. If you're overriding an existing compute and it works, then that's fine by me.

**Comment by HekkiMelody** [comment]
Sorry for my oversights during my first review. Trying to do too many things at once will do that to you.

Adding the field parameters to `sale_discount` seems to be the only thing left. After that I'll approve.

File: sale_order_general_discount_payment_term/models/account_payment_term.py
Comment (HekkiMelody) [inline sale_order_general_discount_payment_term/models/account_payment_term.py]: as you correctly spotted in the other comment, I was looking at v12 code during my previous review.

`track_visibility` was a v12 option, in v14 it's simply `tracking=True`

**Comment by HekkiMelody** [comment]
Code review looks good aside from that one small thing, please squash everything into a single commit.

**Review by HekkiMelody — APPROVED** [review]

**Comment by francesco-ooops** [comment]
@OCA/crm-sales-marketing-maintainers good for merge?

**Comment by francesco-ooops** [comment]
@pedrobaeza could this be merged please?

File: sale_order_general_discount_payment_term/views/res_partner.xml
Comment (pedrobaeza) [inline sale_order_general_discount_payment_term/views/res_partner.xml]: This is not correct. This is out of scope of the module. And there's already a group for hiding this for those you don't want to see it.

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

File: sale_order_general_discount_payment_term/views/res_partner.xml
Comment (francesco-ooops) [reply sale_order_general_discount_payment_term/views/res_partner.xml]: @pedrobaeza the point of this change is that if there is a general discount set in partner and another general discount set in payment term they will be in conflict for setting the general discount on SO. 

How else do you suggest we avoid this scenario?

Comment (pedrobaeza) [reply sale_order_general_discount_payment_term/views/res_partner.xml]: OK, then hide it only if the payment term is set and it contains a discount.

Comment (francesco-ooops) [reply sale_order_general_discount_payment_term/views/res_partner.xml]: Sorry, the flow is unclear to me. What happens if partner 1 has a default general discount set and in SO user selects a payment term with a general discount? Which general discount should be applied?

With this module installed, it's very likely that the only general discount that should be applied is the payment term one. 

Shall we make it easier with a flag in settings allowing user to decide whether to hide the general discount in partner field?

Comment (pedrobaeza) [reply sale_order_general_discount_payment_term/views/res_partner.xml]: You should clear that general discount on payment term assignment IMO.

Comment (pedrobaeza) [reply sale_order_general_discount_payment_term/views/res_partner.xml]: The question is that you shouldn't remove the possibility of assigning directly a general discount for all cases. Just for those covered by this module (with payment term). If not, you will force to assign a payment term always if you want a general discount.

Comment (francesco-ooops) [reply sale_order_general_discount_payment_term/views/res_partner.xml]: User has the possibility of assigning general discount in SO in 3 ways: 

1) default from partner
2) manually from field "discount" in SO
3) default from payment term

base module "sale_order_general_discount" implements the first two options; this module implements the third.

I think if we add a "hide default general discount in partner" flag in settings to this module we will have the cleaner implementation as the user can decide:

- whether to keep all 3 options,
- to keep only 2 and 3 (flag enabled), or 
- to keep 1 and 2 by not installing this module. 

Could that work?

Comment (pedrobaeza) [reply sale_order_general_discount_payment_term/views/res_partner.xml]: I still think that you don't need another extra config. Just don't hide the field if you don't have a payment term selected.

Comment (francesco-ooops) [reply sale_order_general_discount_payment_term/views/res_partner.xml]: Alright, @odooNextev can you implement that?

Comment (francesco-ooops) [reply sale_order_general_discount_payment_term/views/res_partner.xml]: @pedrobaeza with the latest change by @odooNextev , if payment term discount = 0 and partner default discount is set, partner default is used in SO; otherwise payment term discount is used. Could this work?

**Review by pedrobaeza — APPROVED** [review]
/ocabot merge nobump

**Pull request merged**
